firedrake/adjoint/function.py

```diff
@@ -0,0 +1,243 @@
+            """To disable the annotation, just pass :py:data:`annotate=False` to this routine, and it acts exactly like the
+            Dolfin assign call."""
+
+            # do not annotate in case of self assignment
+            annotate = annotate_tape(kwargs) and self != other
+
+            if annotate:
+                if not isinstance(other, ufl.core.operator.Operator):
+                    other = create_overloaded_object(other)
+                block = FunctionAssignBlock(self, other)
+                tape = get_working_tape()
+                tape.add_block(block)
+
+            with stop_annotating():
+                ret = assign(self, other, *args, **kwargs)
+
+            if annotate:
+                block.add_output(self.create_block_variable())
+
+            return ret
+
+        return wrapper
+
+    def _ad_create_checkpoint(self):
+        return self.copy(deepcopy=True)
+
+    @no_annotations
+    def _ad_convert_type(self, value, options=None):
+        from firedrake import Function, TrialFunction, TestFunction, assemble
+
+        options = {} if options is None else options
+        riesz_representation = options.get("riesz_representation", "l2")
+
+        if riesz_representation == "l2":
+            return Function(self.function_space(), val=value)
+
+        elif riesz_representation == "L2":
+            ret = Function(self.function_space())
+            u = TrialFunction(self.function_space())
+            v = TestFunction(self.function_space())
+            M = assemble(firedrake.inner(u, v)*firedrake.dx)
+            firedrake.solve(M, ret, value)
+            return ret
+
+        elif riesz_representation == "H1":
+            ret = Function(self.function_space())
+            u = TrialFunction(self.function_space())
+            v = TestFunction(self.function_space())
+            M = assemble(firedrake.inner(u, v)*firedrake.dx
+                         + firedrake.inner(firedrake.grad(u), firedrake.grad(v))*firedrake.dx)
+            firedrake.solve(M, ret, value)
+            return ret
+
+        elif callable(riesz_representation):
+            return riesz_representation(value)
+
+        else:
+            raise NotImplementedError(
+                "Unknown Riesz representation %s" % riesz_representation)
+
+    def _ad_restore_at_checkpoint(self, checkpoint):
+        return checkpoint
+
+    @no_annotations
+    def adj_update_value(self, value):
+        self.original_block_variable.checkpoint = value._ad_create_checkpoint()
+
+    @no_annotations
+    def _ad_mul(self, other):
+        from firedrake import Function
+
+        r = Function(self.function_space())
+        Function.assign(r, self * other)
+        return r
+
+    @no_annotations
+    def _ad_add(self, other):
+        from firedrake import Function
+
+        r = Function(self.function_space())
+        Function.assign(r, self + other)
+        return r
+
+    def _ad_dot(self, other, options=None):
```

Comment:
`@no_annotations` required?

Reply:
I don't think so, because this only gets called inside pyadjoint and at that stage annotations are guaranteed to be off.